base/commands/migration/start_stages.go

```diff
@@ -0,0 +1,171 @@
+package migration
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	clcerrors "github.com/hazelcast/hazelcast-commandline-client/errors"
+	"github.com/hazelcast/hazelcast-go-client"
+	"github.com/hazelcast/hazelcast-go-client/serialization"
+	"github.com/hazelcast/hazelcast-go-client/types"
+	"golang.org/x/exp/slices"
+
+	"github.com/hazelcast/hazelcast-commandline-client/clc/ux/stage"
+	"github.com/hazelcast/hazelcast-commandline-client/internal/plug"
+)
+
+type StartStages struct {
+	migrationID       string
+	configDir         string
+	ci                *hazelcast.ClientInternal
+	startQueue        *hazelcast.Queue
+	statusMap         *hazelcast.Map
+	updateTopic       *hazelcast.Topic
+	topicListenerID   types.UUID
+	updateMessageChan chan UpdateMessage
+}
+
+var timeoutErr = fmt.Errorf("migration could not be completed: reached timeout while reading status: "+
+	"please ensure that you are using Hazelcast's migration cluster distribution and your DMT config points to that cluster: %w",
+	context.DeadlineExceeded)
+
+func NewStartStages(migrationID, configDir string) *StartStages {
+	if migrationID == "" {
+		panic("migrationID is required")
+	}
+	return &StartStages{
+		migrationID: migrationID,
+		configDir:   configDir,
+	}
+}
+
+func (st *StartStages) Build(ctx context.Context, ec plug.ExecContext) []stage.Stage {
+	return []stage.Stage{
+		{
+			ProgressMsg: "Connecting to the migration cluster",
+			SuccessMsg:  "Connected to the migration cluster",
+			FailureMsg:  "Could not connect to the migration cluster",
+			Func:        st.connectStage(ctx, ec),
+		},
+		{
+			ProgressMsg: "Starting the migration",
+			SuccessMsg:  "Started the migration",
+			FailureMsg:  "Could not start the migration",
+			Func:        st.startStage(ctx, ec),
+		},
+		{
+			ProgressMsg: "Migrating the cluster",
+			SuccessMsg:  "Migrated the cluster",
+			FailureMsg:  "Could not migrate the cluster",
+			Func:        st.migrateStage(ctx, ec),
+		},
+	}
+}
+
+func (st *StartStages) connectStage(ctx context.Context, ec plug.ExecContext) func(stage.Statuser) error {
+	return func(status stage.Statuser) error {
+		var err error
+		st.ci, err = ec.ClientInternal(ctx)
+		if err != nil {
+			return err
+		}
+		st.startQueue, err = st.ci.Client().GetQueue(ctx, StartQueueName)
+		if err != nil {
+			return err
+		}
+		st.statusMap, err = st.ci.Client().GetMap(ctx, MakeStatusMapName(st.migrationID))
+		if err != nil {
+			return err
+		}
+		st.updateTopic, err = st.ci.Client().GetTopic(ctx, MakeUpdateTopicName(st.migrationID))
+		if err != nil {
+			return err
+		}
+		st.updateMessageChan = make(chan UpdateMessage)
+		_, err = st.updateTopic.AddMessageListener(ctx, st.topicListener)
```

Comment:
@kutluhanmetin A generic question: Does go client remove all the subscriptions and stuff upon exit? If I do ctrl+c, I think go client shutdown() will be called by clc code and it could be doing this cleanup. Anyway, I think upon client disconnection, server side should be removing any related runners with that client.